contracts/CreditGuild.sol

```diff
@@ -0,0 +1,177 @@
+            address member = initialMembers[i];
+
+            // check that this member is vouching for the DAO
+            require(userManager.getVouchingAmount(member, address(this)) > 0, "!vouching");
+
+            // pay membership fee
+            underlyingToken.transferFrom(member, address(this), membershipFee);
+
+            // set vouch_amount
+            userManager.updateTrust(member, vouchAmount);
+
+            // mint NFT which is their membership
+            mintNFT(member);
+        } 
+
+        // call stake on behalf of the dao, call register member on behalf of the dao
+
+        emit Initialize(initialMembers);
+
+    }
+
+    function setVouchAmount(uint256 _vouchAmount) public virtual onlyOwner {
+        vouchAmount = _vouchAmount;
+        emit SetVouchAmount(vouchAmount);
+    }
+
+    function setMembershipFee(uint256 _membershipFee) public virtual onlyOwner {
+        membershipFee = _membershipFee;
+        emit SetMembershipFee(membershipFee);
+    }
+
+    // TODO: this wont work right now because of the beforeTokenTransfer override
+    function burnMembership(address member) public virtual onlyOwner {
+        uint256 memberId = addressToId[member];
+        _burn(memberId);
+        emit BurnMembership(member, memberId);
+
+    }
+
+    function checkIsMember(address potentialMember) public virtual returns(bool) {
+      return balanceOf(potentialMember) > 0;
+    }
+
+    function register(address[] memory members) public virtual {
+        require(members.length == 3, "!3 members");
+        require(!checkIsMember(msg.sender), "member");
+
+        for (uint256 i = 0; i < members.length; i++) {
+          require(checkIsMember(members[i]), "!member");
+          require(userManager.getVouchingAmount(members[i], msg.sender) > 0, "!vouching");
+        }
+
+        // Mint the NFT (membership NFT)
+        mintNFT(msg.sender);
+
+        // DAO vouches for this new member
+        userManager.updateTrust(msg.sender, vouchAmount);
+
+        emit Register(msg.sender, members, vouchAmount);
+    }
+
+    // members can call this function to updateTrust on themselves, for example
+    function claimVouch() public {
+
+        userManager.updateTrust(msg.sender, vouchAmount);
+
+        emit ClaimVouch(msg.sender, vouchAmount);
+    }
+
+    function stake() public virtual {
+        // get total DAI
+        uint256 DAOBalance = balanceOf(address(this));
+        // stake all the DAI
+        userManager.stake(DAOBalance);
+    }
+
+    function mintNFT(address member) internal {
+        // mint their NFT - which is their membership
+        uint256 currentId = ++id;
+        _safeMint(member, currentId);
+        addressToId[member] = currentId;
+        emit MintNFT(member, currentId);
+    }
+
+    // calls before every ERC721 call
+    function _beforeTokenTransfer(
+        address from,
+        address,
+        uint256  
+    ) internal override pure {
+      if(from != address(0)) {
```

Comment:
In order to support the `_burn()` function we can update this to allow the `to` address to be `address(0)`

```solidity
    function _beforeTokenTransfer(
        address from,
        address to,
        uint256  
    ) internal override pure {
         require(from == address(0)  || to == address(0), "!transfer");
    }
```